### PR TITLE
Review - 1

### PR DESCRIPTION
**Pull request by arunpatro**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -20,15 +20,16 @@ git clone https://github.com/arunpatro/human-eval
 cd human-eval && pip install -e .
 ```
 
-Checkout the `nbs/humaneval.ipynb` for a demo.
+Checkout the `nbs/humaneval.ipynb` for a demo. on how to use
```
Comment (rmshin) [inline README.md]: for a demo on how to use.

File: web/index.html [hunk 2]
```diff
@@ -63,13 +63,16 @@ <h1>Monte Carlo Tree Search for Code Generation using LLMs</h1>
...
+      even for complex problem descriptions, albeit with some hallucinations due to the probabilistic
       nature of LLMs. In this blog post we explore the following:
       <em
         >is it possible to achieve LLM code generation that is accurate, consistent, and relatively
         cheap to run?</em
       >
     </p>
+    <p>
+      We steer for accuracy by partial scoring in test benches, consistency by using greedy decoding, and cheap by small quantized models.
+    </p>
```
Comment (rmshin) [inline web/index.html]: I think I understand why you added this part, but as it stands I don't feel it reads/flows smoothly with the rest of the section. Don't have the time to suggest a better alternative atm though.

File: web/index.html [hunk 3]
```diff
@@ -89,10 +92,9 @@ <h1>Monte Carlo Tree Search for Code Generation using LLMs</h1>
     </p>
     <h2>Background</h2>
     <p>
-      At a high level, language models output the tokens (characters) that they've been trained to
-      recognise as the most probable ones to come after a given sequence or prompt. It can be seen
-      as a kind of <em>statistical predictor</em>, and it turns out that such an approach works
-      surprisingly well in generating all kinds of text (like source code).
+      Language models function by predicting the most likely next tokens (sequence of characters) that follows a given prompt or sequence. They generate text based on probabilities and patterns derived from their training data, essentialy acting as statistical predictors.
+      It turns out that such an approach works
+      surprisingly well in generating all kinds of text including source code for computer programs.
```
Comment (rmshin) [inline web/index.html:96]: nice rephrasing

File: web/index.html [hunk 4]
```diff
@@ -117,18 +119,17 @@ <h2>Background</h2>
       Example of incorrect Python code output from DeepSeek Coder 6.7B (running on llama.cpp).
     </p>
     <p>
-      There are a number of ways we could approach solving this problem, but the simplest is to just
-      run the LLM multiple times for the same prompt and filter out any incorrect ones that fail
+      There are a number of ways we could mitigate the effect of hallucinations. 
+      We just naively temperature sample the LLM multiple times for the same prompt and filter out the incorrect ones that fail
```
Comment (rmshin) [inline web/index.html]: think I'd prefer this instead:

"There are a number of ways we could mitigate the effect of hallucinations. We could, for instance, just naively sample the LLM multiple times for the same prompt and filter out the incorrect ones that fail..."

File: web/index.html [hunk 5]
```diff
@@ -117,18 +119,17 @@ <h2>Background</h2>
...
       compare alternative solutions.
     </p>
     <p>
-      Another approach is, rather than relying on the LLM to directly generate code on its own, to
-      incorporate its predictions as part of a
+      Another approach is to search for the correct sequence of tokens using a
       <a href="https://en.wikipedia.org/wiki/Tree_(data_structure)" target="_blank">tree-based</a>
       search algorithm. Neither strategy is likely to prove definitive in solving the issue of LLM
-      accuracy, but we'll explore them in greater depth to learn exactly how and when they might be
+      accuracy, simply because the learned probabilities are miscalibrated. We'll explore them in greater depth to learn exactly how and when they might be
```
Comment (rmshin) [inline web/index.html]: can you elaborate on this sentence? Not sure I follow what you're trying to say.

File: web/index.html [hunk 6]
```diff
@@ -193,6 +194,9 @@ <h2>Setting up with an LLM</h2>
         >llama-cpp-python</a
       >.
     </p>
+    <p>
+      We would like to point llama-cpp-python bindings did not yet support multibatch sampling, so we had to run the model in a single batch mode, sequentially. However, due to default prompt kv-caching, there was significant speedup in subsequent runs of the same prompt.
+    </p>
```
Comment (rmshin) [inline web/index.html]: I think it'd be better for this section to come after the code snippet. Also a small phrasing suggestion:

`
We would like to point out that <code class="inline-code">llama-cpp-python</code> does not yet support batched inference, so we had to run the model in a single batch mode sequentially. However, thanks to default prompt kv-caching in <code class="inline-code">llama.cpp</code>, we were still able to process subsequent runs of the same prompt relatively quickly.
`

File: web/index.html [hunk 7]
```diff
@@ -266,9 +270,7 @@ <h2>The HumanEval dataset</h2>
       <a href="https://deepgram.com/learn/humaneval-llm-benchmark#the-passk-metric" target="_blank"
         >pass@k</a
       >
-      metric for the LLM-generated code samples. <b>Pass@k</b> is a probabilistic calculation that
-      represents the likelihood that at least one of the top k-generated samples for each HumanEval
-      task passes its tests. It can be seen as the overall "pass rate" of the LLM over the entire
+      metric for the LLM-generated code samples. The <b>pass@k</b> metric is defined as the probability that at least one of the k-generated samples for each task in the benchmark passes its unit tests. It can be seen as the overall "problem pass rate" of the LLM over the entire
```
Comment (rmshin) [inline web/index.html]: "... It can be seen as the overall "problem pass rate" of the LLM **across** the entire ..."

File: web/index.html [hunk 8]
```diff
@@ -444,15 +446,17 @@ <h2>Adopting MCTS with LLMs</h2>
     </p>
     <p>
       In the <b>selection</b> phase, the MCTS algorithm makes a decision on which node or branch it
-      should explore next by assigning a kind of "priority" to each node. Greater weight is
-      typically given to nodes that are either higher in value or have fewer visits. For our case of
+      should explore or exploit next by assigning a kind of "priority" to each node. Greater weight is
```
Comment (rmshin) [inline web/index.html]: maybe just "expand" instead of "explore or exploit"

File: web/index.html [hunk 9]
```diff
@@ -444,15 +446,17 @@ <h2>Adopting MCTS with LLMs</h2>
...
+      typically given to nodes that are either higher in value (currently observed performance) or have fewer visits (not yet explored i.e. uncertainity). For our case of
```
Comment (rmshin) [inline web/index.html]: "not yet explored i.e. greater uncertainty"

File: web/index.html [hunk 10]
```diff
@@ -444,15 +446,17 @@ <h2>Adopting MCTS with LLMs</h2>
...
       LLM code generation, the exact calculation of this priority is given by:
     </p>
     <img loading="lazy" src="public/p-ucb.png" alt="p-ucb function" class="p-ucb" />
     <p>and the node with the highest priority is selected:</p>
     <img loading="lazy" src="public/p-ucb-select.png" alt="p-ucb selection" class="p-ucb-select" />
     <p>
-      It is not necessary to understand the above in detail. What's more important is to realise
+      This formulation adapts a popular Upper Confidence Bound metric in RL with transfomer probabilities to weight the exploration, see more in <a href="https://arxiv.org/abs/2303.05510" target="_blank"
```
Comment (rmshin) [inline web/index.html]: Suggestion:

`
This formulation adapts the upper confidence bound heuristic with transformer probabilities to weight the exploration, as described in Section D.1 of <a href="https://arxiv.org/abs/2303.05510" target="_blank">Planning with Large Language Models for Code Generation</a>. The choice of which nodes to explore is thus a function of ....
`

File: web/index.html [hunk 11]
```diff
@@ -720,6 +724,7 @@ <h4>Discussions</h4>
         found. But due to the probablistic nature of sampling, which cannot guarantee replication
         across successive generations, we need to collect the full set of samples for every problem
         in order to properly estimate the pass@k metrics for various values of k.
+        One can definitely sample until success and calculate the empiricial expected k to compare with average rollout, but we suffice with estimating the pass@k metric.
```
Comment (rmshin) [inline web/index.html]: We could alternatively sample until success and calculate the empiricial expected k to compare with average rollout, but we suffice with estimating the pass@k metric.

File: web/index.html (hunk 8 above)
Comment (arunpatro) [reply web/index.html]: explore or exploit are good words to introduce here due to RL

File: web/index.html (hunk 5 above)
Comment (arunpatro) [reply web/index.html]: What I meant to say was there is not much what we can do at inference time either with S+F or MCTS if the probabilites are bad. But maybe it is apparent. 

I can remove it but because there is a claim - "Neither strategy is likely to prove definitive" i wanted to add something.

File: web/index.html (hunk 2 above)
Comment (arunpatro) [reply web/index.html]: I rewrote it, let me know if its better in the third para, else we can drop it for lack of time to brainstorm.

File: web/index.html [hunk 12]
```diff
@@ -79,25 +79,23 @@ <h1>Monte Carlo Tree Search for Code Generation using LLMs</h1>
       (MCTS) algorithm to guide the generation, taking inspiration from the paper
       <a href="https://arxiv.org/abs/2303.05510" target="_blank"
         >Planning with Large Language Models for Code Generation</a
-      >
-      . We compare the effectiveness of this approach against direct sampling of the transformer’s
+      >. We steer for <em>accuracy</em> by partial scoring in test benches, <em>consistency</em> by using greedy decoding, and <em>cheap</em> by small quantized models. We compare the effectiveness of this approach against direct sampling of the transformer’s
```
Comment (rmshin) [inline web/index.html]: Hmmm I'm not fully sold on this addition, but if you'd like to keep it I'd maybe suggest something like:

`
We steer for accuracy by continuously scoring partially complete programs on a bench of unit tests, and achieve consistency by using deterministic greedy decoding in the gradual generation of text.
`

`
Relying on a single consumer machine, we compare the effectiveness of this approach against direct sampling of the transformer’s ...
`

File: web/index.html (hunk 5 above)
Comment (rmshin) [reply web/index.html]: yea ok maybe better to write something like this instead then:

"Each strategy comes with its own pros and cons in improving LLM accuracy, and we'll explore them both in greater depth to learn exactly how and when they might be most useful."

File: web/index.html [hunk 13]
```diff
@@ -266,19 +265,17 @@ <h2>The HumanEval dataset</h2>
...
+      metric for the LLM-generated code samples. The <b>pass@k</b> metric is defined as the probability that at least one of the k-generated samples for each task in the benchmark passes its unit tests. It can be seen as the overall "problem pass rate" of the LLM across the entire
       HumanEval problem set.
     </p>
 
     <h2>Generating S+F baselines</h2>
     <p>
       With the S+F strategy, we expected a higher number of samples generated per prompt to increase
       the LLM's accuracy. In an ideal world, we would've generated upwards of ~200 samples, but due
-      to both hardware & time constraints we limited the generations to 20 samples per task. This
+      to both hardware & time constraints, we limited the generations to 20 samples per task. This
```
Comment (rmshin) [inline web/index.html:277]: I think this comma is unnecessary and makes reading less smooth. It's a stylistic point but personally am not a fan of too many commas breaking up a relatively short sentence.

File: web/index.html (hunk 8 above)
Comment (rmshin) [reply web/index.html]: yea we do bring up the explore & exploit terms later on in the post, but fine to mention them here too. I'm nitpicking but in this case can we use "explore/exploit" rather than "explore or exploit"?

File: web/index.html [hunk 14]
```diff
@@ -444,16 +441,17 @@ <h2>Adopting MCTS with LLMs</h2>
...
+      typically given to nodes that are either higher in value (currently observed performance) or have fewer visits (not yet explored i.e. greater uncertainty). For our case of
       LLM code generation, the exact calculation of this priority is given by:
     </p>
     <img loading="lazy" src="public/p-ucb.png" alt="p-ucb function" class="p-ucb" />
     <p>and the node with the highest priority is selected:</p>
     <img loading="lazy" src="public/p-ucb-select.png" alt="p-ucb selection" class="p-ucb-select" />
     <p>
-      It is not necessary to understand the above in detail. What's more important is to realise
-      that the choice of which nodes to explore is a function of three main parameters: the node
+      This formulation adapts a popular Upper Confidence Bound heuristic in RL with transformer probabilities to weight the exploration, as described in Section D.1 of <a href="https://arxiv.org/abs/2303.05510" target="_blank"
```
Comment (rmshin) [inline web/index.html]: I think "...a popular Upper Confidence Bound heuristic in RL..." is redundant here because we mention the UCB heuristic above already with a link to its Wikipedia page. Simply saying "the Upper Confidence Bound heuristic" should be ok.

File: web/index.html [hunk 15]
```diff
@@ -469,23 +467,23 @@ <h2>Adopting MCTS with LLMs</h2>
...
       We achieve this simulation by having an LLM complete the remaining sequence (program) from the
       node's partial state. This completed program is then evaluated against a bench of tests from
       the HumanEval dataset, to obtain a pass rate between 0 - 1 as the node's value. Code that
-      passes more tests end up with higher values, and consequently are explored more in future
+      passes more tests ends up with higher values and consequently is explored more in future
       iterations through the tree.
     </p>
     <p>
-      But due to the nested nature of search trees, any high-value nodes discovered deep in the
+      However due to the nested nature of search trees, any high-value nodes discovered deep in the
```
Comment (rmshin) [inline web/index.html]: We already use "however" earlier above, think it's fine to keep it "but" for this paragraph.

File: web/index.html [hunk 16]
```diff
@@ -495,7 +493,7 @@ <h2>Coding the MCTS algorithm</h2>
       We use a simple <code class="inline-code">Node</code> class to represent the nodes of the
       Monte-Carlo tree. Remember that a node is equivalent to a branch, and it represents a possible
       token (character) that could come next in the sequence of generated text. Each node carries
-      its own values for generation state, number of visits, node value, LLM token probability, and
```
Comment (rmshin) [inline web/index.html:498]: "own" here is important to keep, otherwise the meaning of the sentence subtly changes

File: web/index.html [hunk 17]
```diff
@@ -717,24 +715,25 @@ <h4>Discussions</h4>
...
+        to properly estimate the pass@k metrics for various values of k.
+        We could alternatively sample until success and calculate the empirical expected k to compare with the average rollout of MCTS, but we suffice with estimating the pass@k metric.
       </li>
     </ol>
     <h2>Extending the experiment to VerilogEval</h2>
     <p>
       Using MCTS on the HumanEval dataset proved quite successful, outperforming the ~equivalent S+F
       5-sample baseline by a decent margin and even the 20-sample baseline at much less compute
-      cost. But due to the raw strength of
+      cost. However due to the raw strength of
```
Comment (rmshin) [inline web/index.html]: Again, think it's fine to keep "but" here in favour of "however". I see that grammarly has a specific preference for always using however, it's mostly a stylistic point but not a big fan of it personally

File: web/index.html [hunk 18]
```diff
@@ -750,7 +749,7 @@ <h2>Extending the experiment to VerilogEval</h2>
     <p>
       Because we expected the deepseek LLM to be less capable in Verilog, we needed to adjust a few
       generation parameters before running both S+F and MCTS. We increased the context length & max
-      number of generated tokens to account for the greater verbosity of Verilog code, and also
```
Comment (rmshin) [inline web/index.html:753]: this comma helps to break up the otherwise long sentence, would keep

File: web/index.html [hunk 19]
```diff
@@ -858,7 +857,7 @@ <h2 id="discussion">What have we learned?</h2>
     <h2>Future considerations</h2>
     <p>
       Although MCTS' good performance on HumanEval didn't carry through to the VerilogEval dataset,
-      we believe there is opportunity to make it more robust to fluctuations in the underlying LLM
+      we believe there is an opportunity to make it more robust to fluctuations in the underlying LLM
```
Comment (rmshin) [inline web/index.html]: don't agree with this fix, but could change to "... we believe there are opportunities ..." if you'd prefer?

File: web/index.html (hunk 16 above)
Comment (arunpatro) [reply web/index.html:498]: but its refers to its own

File: web/index.html (hunk 13 above)
Comment (arunpatro) [reply web/index.html:277]: i think you have to keep them, they are clauses

File: web/index.html (hunk 12 above)
Comment (arunpatro) [reply web/index.html]: Let's just drop it?

File: web/index.html (hunk 5 above)
Comment (arunpatro) [reply web/index.html:128]: Just removing this.

File: web/index.html (hunk 8 above)
Comment (arunpatro) [reply web/index.html]: done

File: web/index.html (hunk 5 above)
Comment (arunpatro) [reply web/index.html]: sure

File: web/index.html (hunk 12 above)
Comment (rmshin) [reply web/index.html]: yup fine for me

File: web/index.html (hunk 16 above)
Comment (rmshin) [reply web/index.html:498]: without "own" it doesn't necessarily imply the values are unique to the node, but actually doesn't really matter so all good.
